File: app/connector/factory.py
from app.connector.delay import DelayConnector
from app.connector.enum import ConnectorType
from app.connector.webhook import WebhookConnector

CONNECTORS = [DelayConnector(), WebhookConnector()]


class ConnectorFactory:
    """
    Factory class to create instances of different connectors based on the type.
    """

    @staticmethod
    def get_instance(connector_type: ConnectorType):
        """
        Create an instance of the specified connector type.

        Args:
            connector_type (ConnectorType): The type of connector to create.

        Returns:
            BaseConnector: An instance of the specified connector.

        Raises:
            ValueError: If the connector type is unknown.
        """
        for connector in CONNECTORS:
            if connector.type == connector_type:
                return connector
        raise ValueError(f"Unknown connector type: {connector_type}")

File: rust-app/src/connector/factory.rs
use crate::connector::base::Connector;
use crate::connector::delay::DelayConnector;
use crate::connector::webhook::WebhookConnector;
use crate::schemas::workflow::ConnectorType;

/// Factory for creating connector instances based on type (Factory Pattern).
///
/// Mirrors Python's ConnectorFactory. Uses Open/Closed principle —
/// new connector types can be added without modifying existing code.
pub struct ConnectorFactory;

impl ConnectorFactory {
    /// Get a connector instance for the given type.
    ///
    /// # Arguments
    /// * `connector_type` - The type of connector to create.
    ///
    /// # Returns
    /// A boxed trait object implementing Connector.
    ///
    /// # Errors
    /// Returns an error if the connector type is unknown.
    pub fn get_instance(
        connector_type: &ConnectorType,
    ) -> Result<Box<dyn Connector>, String> {
        match connector_type {
            ConnectorType::Delay => Ok(Box::new(DelayConnector::new())),
            ConnectorType::Webhook => Ok(Box::new(WebhookConnector::new())),
        }
    }
}
